### PR TITLE
ES2015 Lexical this fixes (fat arrows)

### PR DESCRIPTION
**Pull request by nickux**

### CODE REVIEW

**Comment by nickux** [comment]
Tests will pass once #114 is merged and this branch is rebased.

**Comment by jedireza** [comment]
Some of these fat arrow changes are good and others I don't think should use fat arrows even though it works. I'll comment inline on the ones I don't think should be changed.

File: server/models/auth-attempt.js [hunk 1]
```diff
@@ -53,22 +53,20 @@ AuthAttempt.create = function (ip, username, callback) {
 
 AuthAttempt.abuseDetected = function (ip, username, callback) {
 
-    const self = this;
-
     Async.auto({
-        abusiveIpCount: function (done) {
+        abusiveIpCount: (done) => {
```
Comment (jedireza) [inline server/models/auth-attempt.js]: shouldn't be a fat arrow

File: server/models/auth-attempt.js [hunk 2]
```diff
@@ -53,22 +53,20 @@ AuthAttempt.create = function (ip, username, callback) {
...
 
             const query = { ip: ip };
-            self.count(query, done);
+            this.count(query, done);
         },
-        abusiveIpUserCount: function (done) {
+        abusiveIpUserCount: (done) => {
```
Comment (jedireza) [inline server/models/auth-attempt.js]: shouldn't be a fat arrow

File: server/models/session.js [hunk 3]
```diff
@@ -61,11 +61,9 @@ Session.generateKeyHash = function (callback) {
 
 Session.create = function (userId, callback) {
 
-    const self = this;
-
     Async.auto({
         keyHash: this.generateKeyHash.bind(this),
-        newSession: ['keyHash', function (done, results) {
+        newSession: ['keyHash', (done, results) => {
```
Comment (jedireza) [inline server/models/session.js]: shouldn't be a fat arrow

File: server/models/session.js [hunk 4]
```diff
@@ -73,16 +71,16 @@ Session.create = function (userId, callback) {
                 time: new Date()
             };
 
-            self.insertOne(document, done);
+            this.insertOne(document, done);
         }],
-        clean: ['newSession', function (done, results) {
+        clean: ['newSession', (done, results) => {
```
Comment (jedireza) [inline server/models/session.js]: shouldn't be a fat arrow

File: server/models/session.js [hunk 5]
```diff
@@ -99,12 +97,10 @@ Session.create = function (userId, callback) {
 
 Session.findByCredentials = function (id, key, callback) {
 
-    const self = this;
-
     Async.auto({
-        session: function (done) {
+        session: (done) => {
```
Comment (jedireza) [inline server/models/session.js]: shouldn't be a fat arrow

File: server/models/user.js [hunk 6]
```diff
@@ -38,20 +38,19 @@ const User = BaseModel.extend({
             return callback(null, this._roles);
         }
 
-        const self = this;
         const tasks = {};
 
         if (this.roles.account) {
-            tasks.account = function (done) {
+            tasks.account = (done) => {
```
Comment (jedireza) [inline server/models/user.js]: shouldn't be a fat arrow

File: server/models/user.js [hunk 7]
```diff
@@ -38,20 +38,19 @@ const User = BaseModel.extend({
...
 
-                Account.findById(self.roles.account.id, done);
+                Account.findById(this.roles.account.id, done);
             };
         }
 
         if (this.roles.admin) {
-            tasks.admin = function (done) {
+            tasks.admin = (done) => {
```
Comment (jedireza) [inline server/models/user.js]: shouldn't be a fat arrow

File: server/models/user.js [hunk 8]
```diff
@@ -128,11 +127,9 @@ User.generatePasswordHash = function (password, callback) {
 
 User.create = function (username, password, email, callback) {
 
-    const self = this;
-
     Async.auto({
         passwordHash: this.generatePasswordHash.bind(this, password),
-        newUser: ['passwordHash', function (done, results) {
+        newUser: ['passwordHash', (done, results) => {
```
Comment (jedireza) [inline server/models/user.js]: shouldn't be a fat arrow

File: server/models/user.js [hunk 9]
```diff
@@ -159,10 +156,8 @@ User.create = function (username, password, email, callback) {
 
 User.findByCredentials = function (username, password, callback) {
 
-    const self = this;
-
     Async.auto({
-        user: function (done) {
+        user: (done) => {
```
Comment (jedireza) [inline server/models/user.js]: shouldn't be a fat arrow

**Comment by jedireza** [comment]
Fat arrows were introduced into the code base when linting errors were surfaced by `lab`. I tried to only change the functions it called out needing to be fat arrows; which IIRC are inline callbacks.

The fat arrows used in `./setup.js` are there incorrectly. Would you mind adding the keyword function back to those while we're at it?
https://github.com/jedireza/frame/blob/fcdec337b739aed5d1ad1d8b651d5910919c00c5/setup.js

See: http://hapijs.com/styleguide#callback

**Comment by nickux** [comment]
@jedireza Can you elaborate a little on why my change request to use fat arrows is wrong? My understanding of fat arrow functions ([as a long time coffeescript user](http://coffeescript.org/#fat-arrow)) is to address the exact challenge faced in this PR - which is to negate the need to bind `this` to an external variable so that it can be referenced in an inner function (callback or otherwise). I could understand using the `self` method if both `this` contexts (inner and outer) were referenced in the inner function, but I'm not sure I understand why it makes sense to not use them in this case.

[More 'lexical this' and 'fat arrows' reading here](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Functions/Arrow_functions#Lexical_this). 

Also, I did push a commit to update the setup.js file to properly use fat arrows per your request and the hapi styleguide as well.

Looking forward to your response.

Thanks,
Nick

**Comment by jedireza** [comment]
I wouldn't say the changes are "wrong". My goal is to follow the hapi styleguide conventions and unless `lab` calls a function out to use fat arrows I don't.

Also your PR needs some fixing up.

 - The latest commits don't addressed all the `function`/`=>` changes.
 - Looks like we have commits in here that are already on master. You probably need to rebase.

**Comment by jedireza** [comment]
Hit the wrong button, didn't mean to close.

**Comment by nickux** [comment]
@jedireza Ok, makes sense... thanks for explaining your reasons. I just wanted to make sure I didn't have a misunderstanding of the proper use of fat arrows ;)

Ironically, I think rebasing this PR is what caused these weird diff problems. Looking at this locally, I only see the diff of master, but github is showing changes that already exist in master as part of the diff. In any case, I just submitted a new PR (#120) with what I believe are the necessary changes. This PR can be closed.

Thanks,
Nick

**Pull request closed without merging**